### PR TITLE
Move screen space effects into a separate class

### PR DESCRIPTION
**Pull request by BastiaanOlij**
Moving screen space effects into their own effects class. Includes:
- Screen Space Indirect Lighting
- Screen Space Ambient Occlusion
- Screen Space Reflection

SSIL and SSAO do not have stereoscopic support, will deal with this in a separate PR.
SSR stereoscopic support is not working yet, still debugging the issue.
Everything else is working.

### CODE REVIEW

**Comment by BastiaanOlij** [comment]
@Calinou if you have some good test scenes to test these with, can you give this PR a twirl? I think I've got everything working correctly but I've only tested screen space reflection properly so far.

**Comment by Calinou** [comment]
Tested on https://github.com/godotengine/godot/files/7854794/test_ssil_3.zip, with SSAO, SSR and SSIL enabled at the same time.

I confirm this PR works as expected on both clustered and mobile[^1] backends. It also fixes https://github.com/godotengine/godot/issues/62525 even if this PR is rebased against latest `master` locally.

However, this PR doesn't fix https://github.com/godotengine/godot/issues/56724 or https://github.com/godotengine/godot/issues/50151 – SSIL and SSR artifacts are still common when resizing the viewport.

[^1]: Mobile doesn't support SSAO, SSIL or SSR, but it doesn't print errors when enabling those features.

**Comment by BastiaanOlij** [comment]
This PR is pretty much done except for:
- SSAO and SSIL do not have stereoscopic support, this is something we will do in a later PR.
- SSR does have stereoscopic support but currently this doesn't work. I believe the issue is in the code for unprojecting not working correctly.

Setting this to "ready for review" as all the other work can be checked and fixing SSR for stereoscopic can be moved to a separate PR.

**Comment by BastiaanOlij** [comment]
Found a little more I want to do around SSIL and SSAO

**Comment by BastiaanOlij** [comment]
Allocation of buffers for SSAO is now moved from `RendererSceneRenderRD._process_ssao` to `SSEffects::ssao_allocate_buffers`.

With all buffers in our `SSAORenderBuffers` struct and all settings in our `SSAOSettings` struct, also greatly cleaned up passing things around.

**Comment by BastiaanOlij** [comment]
Made the same change to SSIL, pretty happy about the cleanup this gives, nearly all logic around SSIL, SSAO and SSR is now contained in this class. Really the only bit that we can't make modular is the shader code that is in the scene shader.
Other than that, these effects can be pretty much dropped into a renderer.

**Comment by reduz** [comment]
My general thinking is that I am still not sure the array approach makes more sense for compute multiview, it makes code harder to read, harder to test if you don't have multiview and for compute there is zero difference between doing 1 call or two, or three.

And there is zero effort in just creating the individual textures to pass to the effects if you want to run them more than once. I would be much happier if the compute shaders had no idea about what multiview is and this was handled entirely from C++.

**Comment by reduz** [comment]
I think for practical reasons, we really should do away with it everywhere it makes sense. Otherwise think that contributors who don't care about multiview (the vast majority) that want to tinker with shader will have a very hard time testing to see if these work (or just plain ignore that side of the shader) resulting in more breakage for the multiview users, because honestly we cant expect that every change to a shader needs to be tested in multiview.

So IMO, I would really make the rendering code _the least_ multiview centric as possible. At much require the function to operate on array of buffers (separate elements) but not texture arrays.

**Comment by BastiaanOlij** [comment]
@reduz indeed, my main concern was one of performance whether there is overhead in performing the same action twice. However with screen space effects where we're still performing both passes in the same command queue there doesn't seem to be a difference.

I think the only two things we can't get away from when rendering stereo:
1. when using framebuffers (so raster) we need to use multiview, especially when rendering geometry where multiview does take away a lot of overhead compared to running things twice
2. that we do need to do full unprojecting which means storing projection matrices in uniforms as we run out of space in push constants. Right now I'm introducing these where needed but I'm working on changing the way we're handling scene data so we have a uniform with all data available for any shader to use.

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
@BastiaanOlij You have to think of Compute as something that processes little blocks of data on demand. If you do two calls to it (and not barrier the calls) its exactly the same as doing one for both since you just keep the GPU busy doing what it has to do.

**Comment by reduz** [comment]
would still wait on @clayjohn approval since this touches a lot of his code.

**Comment by BastiaanOlij** [comment]
@reduz  ok, I did add the bariers in this case because I'm processing one eye at a time so we don't have to double the sizes of all the buffers. In this case I wonder if it would make much difference seeing plenty is done in parallel already. But maybe reshuffling and making all intermediate buffers have 2 layers would allow us an improvement here?

File: servers/rendering/renderer_rd/environment/gi.h
Comment (clayjohn) [inline servers/rendering/renderer_rd/environment/gi.h:758]: Does this change go with the other commented out VRS changes?

**Review by clayjohn — COMMENTED** [review]
Looks fine to me. I noticed that there are a lot of VRS changes that have been commented out, are those just placeholders for now? Or is it dead code that should be removed?

File: servers/rendering/renderer_rd/environment/gi.h
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/environment/gi.h:758]: No, that was just a mistake of mine on the VRS PR. I really should open a separate PR for this but seeing I'm expecting this to be merged soonish.

Basically without this fix the shader specialisation with half res, multiview and vrs all on wouldn't work. An unlikely combination :)

**Comment by BastiaanOlij** [comment]
> Looks fine to me. I noticed that there are a lot of VRS changes that have been commented out, are those just placeholders for now? Or is it dead code that should be removed?

I've removed these. I think this is ready now

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
